### PR TITLE
Create SAS tokens for blob->* transfers on the fly in the STE

### PR DESCRIPTION
**Pull request by adreed-msft**
Some background:

Previous iterations of this PR have done this creation in the frontend. This would cause problems if a user attempted to resume a transfer of this style, because the SAS has to come in externally in that scenario. Furthermore, this would have caused issues with S2S transfers, as S2S transfers may potentially contain multiple source containers, and UDKs can only create SAS tokens for containers or blobs, **not** the entire service.

Therefore, I moved it all to the STE. It lives under a single-instance struct named UserDelegationAuthenticationManager, which does two things:

1) If a UDK was created, it maps container names to SAS tokens.
2) If a UDK was not created, it creates empty SAS tokens.

Furthermore, rather than just grabbing the URL once from SIP, we grab it every time we make a call to enable refreshing the underlying UDK and user delegated SAS token

### CODE REVIEW

**Comment by adreed-msft** [comment]
Also todo: Make the tests not panic because something about OAuthTokenManager is nil

**Comment by adreed-msft** [comment]
> Also todo: Make the tests not panic because something about OAuthTokenManager is nil

This didn't get caught in manual testing-- But intuition says that OAuthTokenManager didn't get set up in tests for some reason. Maybe some better nil checking or a mimic of some sort will fix it.

File: ste/mgr-JobPartTransferMgr.go [hunk 1]
```diff
@@ -194,6 +208,19 @@ func (jptm *jobPartTransferMgr) Info() TransferInfo {
...
+
+	// make use of UDAM ONLY in the case of blob -> elsewhere, AND the source SAS is not present to start with.
+	fromTo := jptm.FromTo()
+	if fromTo.IsS2S() && fromTo.From() == common.ELocation.Blob() {
+		// Check srcSAS because an S2S transfer requires that the source is either public OR has a SAS token
+		if len(srcSAS) == 0 {
+			// We're going to safely ignore the error here.
+			// Realistically, there will NEVER be an error.
+			// But, if there is, we get "" back. So there's no need to worry, because srcSAS was already empty.
+			srcSAS, _ = jptm.getSrcUserDelegationSAS(src)
```
Comment (adreed-msft) [inline ste/mgr-JobPartTransferMgr.go]: TODO: Consider logging info about our attempt to create a UDK and our attempt to create a User Delegation SAS token

File: ste/userDelegationAuthenticationManager.go [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+func newUserDelegationAuthenticationManager(serviceURL azblob.ServiceURL) (userDelegationAuthenticationManager, error) {
+	authManager := userDelegationAuthenticationManager{}
+	// Create a user delegation
+	authManager.startTime = time.Now()
+	authManager.expiryTime = authManager.startTime.Add(time.Hour * 24 * 7)
+
+	keyInfo := azblob.NewKeyInfo(authManager.startTime, authManager.expiryTime)
+
+	var err error
+	authManager.credential, err = serviceURL.GetUserDelegationCredential(steCtx, keyInfo, nil, nil)
```
Comment (adreed-msft) [inline ste/userDelegationAuthenticationManager.go]: TODO: Consider dropping errors in their current use here, since errors out of UDAM are ignored anyway, and this is a holdover from the original design

File: ste/mgr-JobPartTransferMgr.go [hunk 3]
```diff
@@ -186,6 +185,21 @@ func (jptm *jobPartTransferMgr) GetSourceCompressionType() (common.CompressionTy
 	return common.GetCompressionType(encoding)
 }
 
+func (jptm *jobPartTransferMgr) getSrcUserDelegationSAS(src string) (string, error) {
+	// panics if udam isn't found
+	udam := jptm.jobPartMgr.GetUserDelegationAuthenticationManagerInstance()
+
+	objectURL, err := url.Parse(src)
+
+	if err != nil {
```
Comment (adreed-msft) [inline ste/mgr-JobPartTransferMgr.go]: TODO: add note that if this fails to parse, something has gone terribly, terribly wrong.

File: ste/mgr-JobMgr.go [hunk 4]
```diff
@@ -313,6 +399,21 @@ func (jm *jobMgr) AddJobPart(partNum PartNumber, planFile JobPartPlanFileName, s
...
+
+	// This is actually a really fantastic place to set up our user delegation authentication manager.
+	// This is because we have no awareness of our source root until now,
+	// and two job parts will never have different sourceroots.
+	// Note that setupUDAM checks if UDAM is nil before attempting to create it.
+	if jpm.Plan().FromTo.IsS2S() && jpm.Plan().FromTo.From() == common.ELocation.Blob() && len(jpm.sourceSAS) == 0 {
+		// So now you may be thinking "Can't you copy safely from a public resource?"
+		// The answer is yes.
+		// If a user is copying from a public resource in this case, one of two things will happen:
+		// 1) setupUDAM will fail cleanly, leaving an empty UDAM instance that does nothing.
```
Comment (adreed-msft) [inline ste/mgr-JobMgr.go]: TODO: Consider the implications of a dummy UDAM instance

File: ste/mgr-JobMgr.go [hunk 5]
```diff
@@ -188,10 +192,92 @@ type jobMgr struct {
...
+			LogError: func(str string) { jm.Log(pipeline.LogError, str) },
+			Panic:    jm.Panic,
+			CallerID: "userDelegationAuthenticationManager",
+			Cancel:   jm.Cancel,
+		}
+
+		// Why craft our own credential info here?
+		// the frontend's may be impure, in the case that we're doing OAuth -> SAS
+		// This means that we'll get the OAuthTokenInfo but the credentialType will be anonymous.
+		credInfo := common.CredentialInfo{CredentialType: common.ECredentialType.OAuthToken()}
```
Comment (adreed-msft) [inline ste/mgr-JobMgr.go]: TODO: consider a better way to get an OAuth token in the STE instead of just carrying one from STE

File: ste/mgr-JobMgr.go [hunk 6]
```diff
@@ -188,10 +192,92 @@ type jobMgr struct {
...
+		// no reason for this to fail, ever.
+		common.PanicIfErr(err)
+
+		// get a service-level URL
+		srcBlobURLParts := azblob.NewBlobURLParts(*srcURL)
+		srcBlobURLParts.BlobName = ""
+		srcBlobURLParts.ContainerName = ""
+
+		// create credentials, xfer options
+		credOption := common.CredentialOpOptions{
```
Comment (adreed-msft) [inline ste/mgr-JobMgr.go]: TODO: consider this logging strategy

It only ends up outputting the actual attempted request itself. This works to show success or failure at the creation of a UDK, but it's something that is only understood by the developer at this point.

File: ste/mgr-JobMgr.go [hunk 7]
```diff
@@ -188,10 +192,92 @@ type jobMgr struct {
 
 	// only a single instance of the prompter is needed for all transfers
 	overwritePrompter *overwritePrompter
+	// only a single instance of the user delegation authentication manager is needed for all transfers
+	udam             *userDelegationAuthenticationManager
+	udamCreationLock sync.Mutex // This is a sanity check because it *may* be possible to have AddJobPart called multiple times at once if transfers get scheduled fast enough.
```
Comment (adreed-msft) [inline ste/mgr-JobMgr.go]: TODO: Consider if this is even necessary

File: common/environment.go [hunk 8]
```diff
@@ -190,7 +190,7 @@ func (EnvironmentVariable) CredentialType() EnvironmentVariable {
 func (EnvironmentVariable) DefaultServiceApiVersion() EnvironmentVariable {
 	return EnvironmentVariable{
 		Name:         "AZCOPY_DEFAULT_SERVICE_API_VERSION",
-		DefaultValue: "2018-03-28",
+		DefaultValue: "2018-11-09",
```
Comment (adreed-msft) [inline common/environment.go]: TODO: Consider the implications of adjusting the default service API version

Comment (adreed-msft) [reply common/environment.go]: It's necessary in this case, but I don't know if it'll cause any other problems that need to be fixed before this can ship.

File: cmd/sync.go [hunk 9]
```diff
@@ -479,13 +480,24 @@ func (cca *cookedSyncCmdArgs) process() (err error) {
...
-		if tokenInfo, err := uotm.GetTokenInfo(ctx); err != nil {
-			return err
+	if cca.fromTo.IsS2S() && cca.sourceSAS == "" {
+		if cca.fromTo.From() == common.ELocation.Blob() && cca.credentialInfo.OAuthTokenInfo != (common.OAuthTokenInfo{}) {
+			glcm.Info("A SAS token is required as a part of the source in S2S sync, unless the source is a public resource. " +
+				"Because no SAS token is present on your blob source, we'll attempt to generate a SAS token at transfer time. Please note that " +
+				"transfers may fail if generating the user delegation SAS token fails.")
 		} else {
-			cca.credentialInfo.OAuthTokenInfo = *tokenInfo
+			return errors.New("a SAS token is required as a part of the source in S2S sync, unless the source is a public resource (or a blob source and you have storage blob data owner permissions)")
```
Comment (adreed-msft) [inline cmd/sync.go]: TODO: Note that SBDO is not actually required here (https://docs.microsoft.com/en-us/rest/api/storageservices/get-user-delegation-key)

File: ste/userDelegationAuthenticationManager.go (hunk 2 above)
Comment (adreed-msft) [reply ste/userDelegationAuthenticationManager.go]: Also TODO: Consider refreshing the UDK, in case customers have long-running scenarios

Comment (adreed-msft) [reply ste/userDelegationAuthenticationManager.go]: The latter is actually a humongous pain to do--  STE may have already taken expired SAS tokens if we don't refresh

A workaround is to just keep resuming the job whenever the token expires.

Comment (adreed-msft) [reply ste/userDelegationAuthenticationManager.go]: WRT to this-- Perhaps passing around pointers to atomic.Values (and ensuring validity before we use a SAS token) would be the better solution in terms of handling SAS tokens in the STE if we need to refresh this.

Comment (adreed-msft) [reply ste/userDelegationAuthenticationManager.go]: Although, thinking about it, this isn't a huge concern-- If somebody somehow runs into this, the suggestion is literally just "create a SAS token that lasts longer"

**Comment by adreed-msft** [comment]
Speaking of which, this'll resolve #211

File: ste/userDelegationAuthenticationManager.go (hunk 2 above)
Comment (adreed-msft) [reply ste/userDelegationAuthenticationManager.go]: Deciding NOT to handle things this way. As-is, despite the lack of use of errors, there _may_ be a use-case in the future.

File: ste/mgr-JobMgr.go (hunk 7 above)
Comment (adreed-msft) [reply ste/mgr-JobMgr.go]: The answer is: not really, but it's still good to have this.

File: common/environment.go (hunk 8 above)
Comment (adreed-msft) [reply common/environment.go]: It seems fine to me, there are no changes to existing APIs, just additions.

File: ste/mgr-JobMgr.go (hunk 5 above)
Comment (adreed-msft) [reply ste/mgr-JobMgr.go]: Added a note here with this TODO and created a PBI. This isn't exactly what this PR is about, anyway.

File: ste/mgr-JobMgr.go (hunk 4 above)
Comment (adreed-msft) [reply ste/mgr-JobMgr.go]: Dummy UDAM instances are _ok_, but we should add some extra logging and stdout warnings for it.

**Comment by adreed-msft** [comment]
Force-pushing to ensure PR is not stale

File: cmd/credentialUtil.go [hunk 10]
```diff
@@ -286,14 +286,15 @@ func getCredentialInfoForLocation(ctx context.Context, location common.Location,
...
 
-	if credInfo.CredentialType == common.ECredentialType.OAuthToken() {
-		uotm := GetUserOAuthTokenManagerInstance()
-
-		if tokenInfo, err := uotm.GetTokenInfo(ctx); err != nil {
-			return credInfo, false, err
-		} else {
-			credInfo.OAuthTokenInfo = *tokenInfo
-		}
+	// Get the OAuth token info anyway. This may be used later in STE.
```
Comment (JohnRusk) [inline cmd/credentialUtil.go]: This change doesn't smell right. Is it necessary?

If user has specifically told us on the command line NOT to use OAuth, why would we try to look up a token?

File: common/environment.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (adreed-msft) [reply common/environment.go]: Nope. The transfers will fail as the user delegation SAS tokens can't be used on the earlier revision.

[the remark this answers is not in the document]
Comment (adreed-msft) [reply common/environment.go]: We could however, just bump it for this type of transfer.

File: ste/mgr-JobMgr.go [hunk 11]
```diff
@@ -188,10 +192,98 @@ type jobMgr struct {
...
+				RetryDelay:    UploadRetryDelay,
+				MaxRetryDelay: UploadMaxRetryDelay},
+			JobsAdmin.(*jobsAdmin).pacer,
+			jm.HttpClient(),
+			nil)
+
+		// finally, we have a service URL.
+		bsu := azblob.NewServiceURL(srcBlobURLParts.URL(), p)
+
+		// Ignoring the error. If we can't create UDAM, we don't have adequate permissions in the first place.
```
[the remark this answers is not in the document]
Comment (adreed-msft) [reply ste/mgr-JobMgr.go]: This was a comment made prior to the err check just below. The error isn't really handled, the failure is just logged.

File: ste/mgr-JobPartTransferMgr.go [hunk 12]
```diff
@@ -185,6 +186,25 @@ func (jptm *jobPartTransferMgr) GetSourceCompressionType() (common.CompressionTy
 	return common.GetCompressionType(encoding)
 }
 
+func (jptm *jobPartTransferMgr) getSrcUserDelegationSAS(src string) (string, error) {
```
[the remark this answers is not in the document]
Comment (adreed-msft) [reply ste/mgr-JobPartTransferMgr.go]: Yup, that came from the old implementation, prior to the usage of SIP

File: ste/userDelegationAuthenticationManager.go [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+		sas, err := u.createUserDelegationSASForURL(blobURLParts.ContainerName)
+		return sas, err
+	}
+}
+
+func (u *userDelegationAuthenticationManager) createUserDelegationSASForURL(containerName string) (string, error) {
+	// If it's not present, we need to generate a SAS query and store it, then return.
+	sasQuery, err := azblob.BlobSASSignatureValues{
+		Version:       DefaultServiceApiVersion,
+		Protocol:      azblob.SASProtocolHTTPSandHTTP, // A user may be inclined to use HTTP for one reason or another. We already warn them about this.
```
[the remark this answers is not in the document]
Comment (JohnRusk) [reply ste/userDelegationAuthenticationManager.go]: I unresolved it because the latest version has the constant changed (great) but still the old comment. Please fix/remove old comment at the end of the line. It only confuses readers if we have code that does one thing and comment that does another.

File: ste/mgr-JobMgr.go [hunk 14]
```diff
@@ -313,6 +430,30 @@ func (jm *jobMgr) AddJobPart(partNum PartNumber, planFile JobPartPlanFileName, s
...
+		jm.initState = &jobMgrInitState{}
+
+		// For UDAM to be initialized, there are some conditions:
+		// 1) The source must not be public (It's unsafe to send the OAuth token to a public source)
+		// 2) It should be a S2S transfer from blob (There's no need for a SAS on a download, and user delegation doesn't work on files)
+		// 3) The source should lack a SAS token (There's no need for SAS generation when we already have)
+		if !jpm.Plan().SourcePublic && jpm.Plan().FromTo.IsS2S() && jpm.Plan().FromTo.From() == common.ELocation.Blob() && len(jpm.sourceSAS) == 0 {
+			// Upgrade the default version for this transfer.
+			// Note that S2S transfers with a User delegation SAS will fail on a lower revision.
+			DefaultServiceApiVersion = "2018-11-09"
```
Comment (JohnRusk) [inline ste/mgr-JobMgr.go]: >DefaultServiceApiVersion = "2018-11-09" [](start = 3, length = 39)

Oh, we need the more recent service version not to create the user delegation SAS, but to USE it?  Have I understood that correctly.  If so, maybe my earlier suggestion not to change the defualt globally was a bad idea.  I think I'd perfer a global change as you origienally did it to changing it here like this.

File: common/LFUCache.go [hunk 15]
```diff
@@ -0,0 +1,147 @@
+package common
+
+import (
+	"sync"
+)
+
+// LFUCache implements a thread-safe least frequently used map cache.
```
[the remark this answers is not in the document]
Comment (JohnRusk) [reply common/LFUCache.go]: This file and its related tests, are on longer used, right? If that's correct, please remove them from the PR.

File: ste/mgr-JobMgr.go [hunk 16]
```diff
@@ -188,10 +198,117 @@ type jobMgr struct {
...
+		}
+
+		srcURL, err := url.Parse(src)
+
+		// no reason for this to fail, ever.
+		common.PanicIfErr(err)
+
+		// Check if the source is pointing at Azure. We shouldn't send the OAuth token anywhere else.
+		// format is account.[blob.core.windows.net], [] is the selection
+		if srcURL.Host[strings.Index(srcURL.Host, ".")+1:] != "blob.core.windows.net" {
```
Comment (JohnRusk) [inline ste/mgr-JobMgr.go]: Not sure we can use this.  Since it doesn't work for sovereign clouds. E.g. For Azure China the accounts end in .blob.core.chinacloudapi.cn

**Comment by antonipp** [comment]
Hi, has this PR been abandoned? This is a very useful feature for us that we'd definitely like to see in `azcopy`.

**Comment by JohnRusk** [comment]
@zezha-msft for comment

**Comment by zezha-msft** [comment]
Thanks @ antonipp for the feedbacks!

This PR has unfortunately fallen out of sync with the underlying dev branch now, so it needs to be rebased and reworked. @adreed-msft will take care of it when her bandwidth frees up.

**Comment by adreed-msft** [comment]
Closing this because the PR has been refreshed: #1375
